File: 1055_shortest_way_to_form_string.py
class Solution:
    def shortestWay(self, source: str, target: str) -> int:
        """
            https://leetcode.com/problems/shortest-way-to-form-string/
            Time Complexity - O(mn)
            'm' is the length of source string
            'n' is the length of target string
            Space Complexity - O(1)
        """
        count = 0
        t_ptr = 0
        while t_ptr < len(target):
            s_ptr = 0
            cur_target = t_ptr
            while s_ptr < len(source) and t_ptr < len(target):
                if source[s_ptr] == target[t_ptr]:
                    s_ptr += 1
                    t_ptr += 1
                else:
                    s_ptr += 1
            if cur_target == t_ptr:
                return -1
            count += 1
        return count


if __name__ == '__main__':
    print(Solution().shortestWay("abc", "abcbc"))
    print(Solution().shortestWay("abc", "acdbc"))

File: 1055_Shortest_Way_To_Form_String.java
    /*  Explanation
    # Leetcode problem link : https://leetcode.com/problems/shortest-way-to-form-string/
    Time Complexity for operators : o(n*m) .. n is the length of the array
    Extra Space Complexity for operators : o(1) 
    Did this code successfully run on Leetcode : NA
    Any problem you faced while coding this : No
# Your code here along with comments explaining your approach
        # Basic approach : 
        # Optimized approach: 
                              
            # 1. 
                    A) 
    */ 

class Solution {
    public int shortestWay(String source, String target) {
        
        int s = source.length();
        int t = target.length();
        
        int TP = 0;
        int SP = 0;
        int output = 0;
        
        while(TP < t){
            SP = 0;
            int curr = TP;
            
            while(TP < t && SP < s){
                if(source.charAt(SP) == target.charAt(TP)){
                    
                    TP += 1;
                }
                SP += 1;
            }
            
            if(curr == TP){
                return -1;
            }
            
            output += 1;
        }
        return output;
    }
}